extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java

```diff
@@ -17,15 +17,46 @@
 
 package org.pentaho.platform.web.http.api.resources.services;
 
+import org.apache.commons.lang3.StringUtils;
 import org.pentaho.platform.api.engine.security.userroledao.IUserRoleDao;
 import org.pentaho.platform.engine.core.system.PentahoSessionHolder;
 import org.pentaho.platform.engine.core.system.PentahoSystem;
+import org.pentaho.platform.web.http.api.resources.UnauthorizedException;
+import org.pentaho.platform.web.http.api.resources.User;
 import org.pentaho.platform.web.http.api.resources.UserListWrapper;
 
 public class UserRoleDaoService {
-    public UserListWrapper getUsers() throws Exception {
-        IUserRoleDao roleDao =
-                PentahoSystem.get( IUserRoleDao.class, "userRoleDaoProxy", PentahoSessionHolder.getSession() );
-        return new UserListWrapper( roleDao.getUsers() );
+  public UserListWrapper getUsers() throws Exception {
+    IUserRoleDao roleDao =
+        PentahoSystem.get( IUserRoleDao.class, "userRoleDaoProxy", PentahoSessionHolder.getSession() );
+    return new UserListWrapper( roleDao.getUsers() );
+  }
+
+  private boolean containsReservedChars( String username ) {
+    StringBuffer reservedChars = new FileService().doGetReservedChars();
+    return StringUtils.containsAny( username, reservedChars );
+  }
+
+  private boolean userValid( User user ) {
+    String name = user.getUserName();
+    String pass = user.getPassword();
+
+    boolean nameValid = ( name != null && name.length() > 0 && !containsReservedChars( name ) );
+    boolean passValid = ( pass != null && pass.length() > 0 );
+    return nameValid && passValid;
+  }
+
+  public void createUser( User user ) throws UnauthorizedException, Exception {
+    if ( userValid( user ) ) {
+      IUserRoleDao roleDao =
+          PentahoSystem.get( IUserRoleDao.class, "userRoleDaoProxy", PentahoSessionHolder.getSession() );
+      roleDao.createUser( null, user.getUserName(), user.getPassword(), "", new String[0] );
+    } else {
+      throw new ValidationFailedException();
```

Comment:
If we are going to throw a validation exception failure here we should also catch this in the resource layer and map it to a response code, to let the user know that the username contains invalid characters. I like this exception for validation though.